### PR TITLE
Enhance MySQL protocol analyzer with new features and fixes - v2

### LINKED ISSUE #2716: MySQL analyzer: Support plugin auth and auth switch
Splitting this out from #2696 - there's a authentication plugin mechanism in the MySQL protocol that we're not properly supporting currently.

The pcap below was taken from a modern MySQL version and a client authenticating with. They negotiate the "caching_sha2_password"  auth plugin, then exchange Auth Switch Request and Responses before switching into the command phase.

That isn't something the current MySQL analyzer supports. It's also depending on capability negotiation between server and client.

[mysql-show-engine-innodb-status-with-password.pcap.zip](https://github.com/zeek/zeek/files/10491546/mysql-show-engine-innodb-status-with-password.pcap.zip)

References:
https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_packets_protocol_handshake_v10.html
https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_initial_handshake_plain_handshake

### PR DESCRIPTION
**Pull request by awelzel**
Follow up for #3803 

Closes #2716

### CODE REVIEW

File: src/analyzer/protocol/mysql/mysql-protocol.pac [hunk 1]
```diff
@@ -173,6 +173,85 @@ enum Client_Capabilities {
...
+
+type BinaryDate = record {
+	len: uint8 &enforce(len == 0 || len == 4 || len == 7 || len == 11);
+	not_implemented: bytestring &length=len;
+};
+
+type BinaryTime = record {
+	len: uint8 &enforce(len == 0 || len == 8 || len == 12);
+	not_implemented: bytestring &length=len;
+};
```
Comment (AmazingPP) [inline src/analyzer/protocol/mysql/mysql-protocol.pac:220]:
```suggestion
type Date = record {
	year : int16;
	month: int8;
	day  : int8;
};

type Time = record {
	hour  : int8;
	minute: int8;
	second: int8;
};

type BinaryDate = record {
	len: uint8 &enforce(len == 0 || len == 4 || len == 7 || len == 11);
	have_date: case ( len > 0 ) of {
		true  -> date  : Date;
		false -> none_1: empty;
	};
	have_time: case ( len > 4 ) of {
		true  -> time  : Time;
		false -> none_2: empty;
	};
	have_micros: case ( len > 7 ) of {
		true  -> micros: int32;
		false -> none_3: empty;
	};
};

type DurationTime = record {
	is_negative: int8 &enforce(is_negative == 0 || is_negative == 1);
	days       : int32;
	time       : Time;
};

type BinaryTime = record {
	len: uint8 &enforce(len == 0 || len == 8 || len == 12);
	have_time: case ( len > 0 ) of {
		true  -> time  : DurationTime;
		false -> none_1: empty;
	};
	have_micros: case ( len > 8 ) of {
		true  -> micros: int32;
		false -> none_2: empty;
	};
};
```

**Review by AmazingPP — CHANGES_REQUESTED** [review]
The additions in this PR are excellent. I noticed that the `BinaryTime` and `BinaryDate` records were not implemented in the PR, so I have written my own implementations for these records. Great work overall!

File: src/analyzer/protocol/mysql/mysql-protocol.pac (hunk 1 above)
Comment (awelzel) [reply src/analyzer/protocol/mysql/mysql-protocol.pac:220]: I'll take it :-) Thanks!

**Review by AmazingPP — APPROVED** [review]
LGTM

File: src/analyzer/protocol/mysql/mysql-protocol.pac [hunk 2]
```diff
@@ -240,13 +240,14 @@ type Handshake_v10 = record {
...
-	reserved               : padding[10]; 
-	auth_plugin_data_part_2: bytestring &length=13;
+	reserved               : padding[10];
+	auth_plugin_data_part_2: bytestring &length=auth_plugin_data_part_2_len;
 	have_plugin : case ( ( capability_flags_2 << 16 ) & CLIENT_PLUGIN_AUTH ) of {
 		CLIENT_PLUGIN_AUTH -> auth_plugin: NUL_String;
 		0x0 -> none    : empty;
 	};
 } &let {
+	auth_plugin_data_part_2_len = (auth_plugin_data_len > 8 && (auth_plugin_data_len - 8) > 13) ? auth_plugin_data_len - 8 : 13;
```
Comment (timwoj) [inline src/analyzer/protocol/mysql/mysql-protocol.pac:356]: Wouldn't this be easier as `auth_plugin_data_len > 21` and a comment for the magic number?

File: src/analyzer/protocol/mysql/mysql-protocol.pac [hunk 3]
```diff
@@ -688,21 +683,16 @@ refine connection MySQL_Conn += {
...
-		return auth_plugin_;
-		%}
-
 	function set_auth_plugin(a: bytestring): bool
 		%{
-		if ( auth_plugin_.length() > 0 &&
-		     strncmp(c_str(auth_plugin_), c_str(a), auth_plugin_.length()) != 0 )
+		// binpac::std_str() includes trailing \0 from parsing.
+		auto new_auth_plugin = std::string(binpac::c_str(a));
+		if ( auth_plugin_.length() > 0 && new_auth_plugin != auth_plugin_ )
```
Comment (timwoj) [inline src/analyzer/protocol/mysql/mysql-protocol.pac:786]: I usually prefer to use `.empty()` vs comparing to zero.

**Review by timwoj — APPROVED** [review]
Just a couple of minor nits. I didn't compare it to the protocol spec at all.

**Pull request merged**